src/placement/jump_map.c

```diff
@@ -336,22 +336,19 @@ uint32_t
 count_available_spares(struct pl_jump_map *jmap, struct pl_obj_layout *layout,
 		uint32_t failed_in_layout)
 {
-	uint32_t num_failed;
-	uint32_t spares_left;
+	uint32_t unusable_tgts;
+	uint32_t num_targets;
 
-	spares_left = jmap->jmp_domain_nr;
-	num_failed = pool_map_get_failed_cnt(jmap->jmp_map.pl_poolmap,
-			jmap->min_redundant_dom);
+	num_targets =  pool_map_find_target(jmap->jmp_map.pl_poolmap,
+			 PO_COMP_ID_ALL, NULL);
 
-	if (spares_left + failed_in_layout < (num_failed + layout->ol_nr))
+	/* we might not have any valid targets left at all */
+	unusable_tgts = layout->ol_nr;
+
+	if (unusable_tgts >= num_targets || layout->ol_grp_size == 1)
 		return 0;
 
-	/* Add back the ones already counted as failed in the layout
-	 * Or we would double count them.
-	 */
-	spares_left = spares_left - (num_failed + layout->ol_nr);
-	spares_left += failed_in_layout;
-	return spares_left;
+	return num_targets - unusable_tgts;
```

Comment:
It looks to me that the old code checks if there is available domain (node), this patch changed it to check if there is available target (VOS)? So we do allow two replicas on same node now?

Reply:
Correct, this should have been find node.

This may need to change in the future if we have object classes of type RP_2SX and RP_3SX.

Reply:
I guess I read the code wrong, 'num_targets' here is the number of nodes in pool map but not number of VOS targets in pool map, right?